Add: doc page for the object detection task (#20925)
huggingface/transformers: docs/source/en/tasks/object_detection.mdx

<!--Copyright 2022 The HuggingFace Team. All rights reserved.

Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
the License. You may obtain a copy of the License at

http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# Object detection

[[open-in-colab]]

Object detection is a computer vision task of detecting instances of semantic objects that belong to a particular
class (such as humans, buildings, or cars) in parts of an image. Object detection models receive an image as input and output
coordinates of the bounding boxes and associated labels of the detected objects. An image can contain multiple objects,

> each with its own bounding box and a label. Similarly, several instances of the same class may

Reviewer: ```suggestion
each with its own bounding box and a label (e.g. it can have a car and a building).
```

be present in different parts of an image.
This task is commonly used in autonomous driving for detecting things like pedestrians, road signs, and traffic lights.
Other applications include counting objects in images, image search, and more.

 <Tip>
 Check out the [object detection task page](https://huggingface.co/tasks/object-detection) to learn about use cases,
 models, metrics, and datasets associated with this task.

>  </Tip>

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

>  </Tip>

Reviewer 2: For Markdown, you need blank lines after `<Tip`> and before `</Tip`> to render the link correctly.

In this guide, you will learn how to:

 1. Finetune [DETR](https://huggingface.co/docs/transformers/model_doc/detr), a model that combines a convolutional
 backbone with an encoder-decoder Transformer, on the [CPPE-5](https://huggingface.co/datasets/cppe-5)
 dataset.
 2. Use your finetuned model for inference.

Before you begin, make sure you have all the necessary libraries installed:

```bash

> pip install -q datasets transformers evaluate timm albumentations opencv-python

Author: Good point. I also realized that I don't need opencv-python, it's a leftover dependency from my earlier experiments, and I don't need it in the current version. 
`timm` is required now, but won't be needed as soon as this https://github.com/huggingface/transformers/pull/20833 is merged

> pip install -q datasets transformers evaluate timm albumentations opencv-python

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

When prompted, enter your token to log in:

```py
>>> from huggingface_hub import notebook_login

>>> notebook_login()
```

## Load the CPPE-5 dataset

In this guide, you'll use the [CPPE-5 dataset](https://huggingface.co/datasets/cppe-5) that contains images with
annotations identifying medical personal protective equipment (PPE) in the context of the COVID-19 pandemic.

Start by loading the dataset:

```py
>>> from datasets import load_dataset

>>> cppe5 = load_dataset("cppe-5")
>>> cppe5
```

You'll see that this dataset already comes with a training set containing 1000 images and a test set with 29 images.

To get familiar with the data, explore what the examples look like.

```py
>>> cppe5["train"][0]
```

The examples in the dataset have the following fields:
- `image_id`: the example image id
- `image`: a `PIL.Image.Image` object containing the image
- `width`: width of the image
- `height`: height of the image
- `objects`: A dictionary containing bounding box metadata for the objects in the image:
  - `id`: The annotation id.
  - `area`: The area of the bounding box.

>   - `bbox`: The object's bounding box (in the [COCO](https://albumentations.ai/docs/getting_started/bounding_boxes_augmentation/#coco) format).

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

  - `category`: The object's category, with possible values including `Coverall (0)`, `Face_Shield (1)`, `Gloves (2)`, `Goggles (3)` and `Mask (4)`.

You may notice that the `bbox` field follows the COCO format, which is the format that DETR model expects.
However, the way grouping of the fields inside `objects` differs from the annotations format DETR requires. You will
need to apply some preprocessing transformations before using this data for training.

To get an even better understanding of the data, visualize a random example in the dataset.

```py
>>> import numpy as np
>>> import os
>>> from PIL import Image, ImageDraw

>>> # based on https://github.com/woctezuma/finetune-detr/blob/master/finetune_detr.ipynb

> >>> image_ids = cppe5["train"]["image_id"]

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

>>> # let's pick a random image
>>> image_id = image_ids[np.random.randint(0, len(image_ids))]
>>> print("Image n°{}".format(image_id))

>>> image = cppe5["train"][image_id]["image"]
>>> annotations = cppe5["train"][image_id]["objects"]

> >>> draw = ImageDraw.Draw(image, "RGBA")

Author: Hm, not really, I can remove it.

> >>> draw = ImageDraw.Draw(image, "RGBA")

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

>>> # Retrieve the categories from the dataset's metadata.
>>> categories = cppe5["train"].features["objects"].feature["category"].names

>>> id2label = {index: x for index, x in enumerate(categories, start=0)}
>>> label2id = {v: k for k, v in id2label.items()}

>>> for i in range(len(annotations["id"])):
...     box = annotations["bbox"][i - 1]
...     class_idx = annotations["category"][i - 1]
...     x, y, w, h = tuple(box)
...     draw.rectangle((x, y, x + w, y + h), outline="red", width=1)
...     draw.text((x, y), id2label[class_idx], fill="white")

> >>> image

Author: Makes sense, I'll create a separate PR to add images and meanwhile I'll add it to the page as an imgur-hosted image.

> >>> image

Author: I added an image in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7
Creating a PR to (documentation-images) is on my TODO.

```

To visualize the bounding boxes with associated labels, you can get the labels from the dataset's metadata, specifically
the `category` field.
You'll also want to create dictionaries that map a label id to a label class (`id2label`) and the other way around (`label2id`).

> You'll need them later when setting up your model.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

As a final step of getting familiar with the data, explore it for potential issues. One common problem with datasets for
object detection is bounding boxes that "stretch" beyond the edge of the image. Such "runaway" bounding boxes can raise
errors during training and should be addressed at this stage. There are a few examples with this issue in this dataset.
To keep things simple, remove these images from the data.

```py
>>> remove_idx = [590, 821, 822, 875, 876, 878, 879]
>>> keep = [i for i in range(len(cppe5["train"])) if i not in remove_idx]
>>> cppe5["train"] = cppe5["train"].select(keep)
```

## Preprocess the data

To finetune a model, you must preprocess the data you plan to use to match precisely the approach used for the pre-trained model.

> For DETR models, `DetrImageProcessor` takes care of processing image data to create `pixel_values`, `pixel_mask`, and

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

`labels` that a DETR model can train with. The preprocessor has some attributes that you won't have to worry about:

- `image_mean = [0.485, 0.456, 0.406 ]`
- `image_std = [0.229, 0.224, 0.225]`

These are the mean and standard deviation used to normalize images during the model training. These values are crucial
to replicate when doing inference or finetuning a pre-trained image model.

Instantiate the `DetrImageProcessor` from the same checkpoint as the model you want to finetune.

```py
>>> from transformers import DetrImageProcessor

>>> checkpoint = "facebook/detr-resnet-50"

> >>> image_processor = DetrImageProcessor.from_pretrained(checkpoint)

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

```

Before passing the images to the `DetrImageProcessor`, apply two preprocessing transformations to the dataset:
- Augmenting images
- Re-formatting annotations to meet DETR expectations

First, to make sure the model does not overfit on the training data, you can apply image augmentation with [Albumentations](https://albumentations.ai/docs/).
This library ensures that transformations affect the image and update the bounding boxes accordingly.
The 🤗 Datasets library documentation has a detailed [guide on how to augment images for object detection](https://huggingface.co/docs/datasets/object_detection),
and it uses the exact same dataset as an example. Apply the same approach here, i.e., resize each image to (480, 480),
flip it horizontally, and brighten it:

```py

> >>> import albumentations as A

Author: This is an example that I took from the [Dataset docs](https://huggingface.co/docs/datasets/v2.8.0/en/object_detection). If I update it here, I should probably create an additional small PR to update it there too for consistency.

> >>> import albumentations as A

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7 for this doc, will create a separate PR for the Dataset docs.

> >>> import albumentations as A

Author: PR in the dataset docs: https://github.com/huggingface/datasets/pull/5403

...         A.RandomBrightnessContrast(p=1.0),
...     ],
...     bbox_params=A.BboxParams(format="coco", label_fields=["category"]),
... )
```

`DetrImageProcessor` expects the annotations to be in the following format: `{‘image_id’: int, ‘annotations’: List[Dict]}`,
 where each `Dict` is a COCO object annotation. Building up gradually, add a function to reformat annotations for a single example:

```py
>>> def formatted_anns(image_id, category, area, bbox):

...     annotations = []
...     for i in range(0, len(category)):
...         new_ann = {
...             "image_id": image_id,
...             "category_id": category[i],
...             "isCrowd": 0,
...             "area": area[i],
...             "bbox": list(bbox[i]),
...         }
...         annotations.append(new_ann)

...     return annotations
```

Now you can combine the image and annotation transformations to use on a batch of examples:

```py
>>> # transforming a batch
>>> def transform_aug_ann(examples):
...     image_ids = examples["image_id"]
...     images, bboxes, area, categories = [], [], [], []
...     for image, objects in zip(examples["image"], examples["objects"]):

> ...         image = np.array(image.convert("RGB"))[:, :, ::-1]

Author: It's RGB. There was an issue in the data where some images were not in RGB. This seems to help.

> ...         image = np.array(image.convert("RGB"))[:, :, ::-1]

Reviewer 3: This is indeed weird. 

`image = np.array(image.convert("RGB"))[:, :, ::-1]` is reversing the order of the channels, isn't it?

> ...         image = np.array(image.convert("RGB"))[:, :, ::-1]

Reviewer 4: Yes, fair point, not sure why BGR is used

> ...         image = np.array(image.convert("RGB"))[:, :, ::-1]

Author: Forgot to mention that it was removed in the example.

...         out = transform(image=image, bboxes=objects["bbox"], category=objects["category"])

...         area.append(objects["area"])

> ...         images.append(torch.tensor(out["image"]).flip(-1).permute(2, 0, 1))

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

...         bboxes.append(out["bboxes"])
...         categories.append(out["category"])

...     targets = [
...         {"image_id": id_, "annotations": formatted_anns(id_, cat_, ar_, box_)}
...         for id_, cat_, ar_, box_ in zip(image_ids, categories, area, bboxes)
...     ]

...     return image_processor(images=images, annotations=targets, return_tensors="pt")
```

Apply this preprocessing function to the entire dataset using 🤗 Datasets `with_transform` method. This method applies
transformations on the fly when you load an element of the dataset.

At this point, you can check what an example from the dataset looks like after the transformations. You should see a tensor
with `pixel_values`, a tensor with `pixel_mask`, and `labels`.

```py
>>> cppe5["train"] = cppe5["train"].with_transform(transform_aug_ann)

> >>> cppe5["train"][15]

Reviewer 3: Feel free to also append the output in a trimmed manner here (c.f. https://moon-ci-docs.huggingface.co/docs/transformers/pr_20925/en/tasks/semantic_segmentation).

Reviewer 3: Might also be worth visualizing how augmented images with bounding boxes look like. 

We might want to reuse the visualization code here. Maybe turn it into a function so that the overall code stays cleaner?

Author: I'll add the output, but I don't think it is necessary to add the visualization. This guide is already quite large, and there's a separate doc page about this exact augmentation on this exact dataset that has all of the visualizations. I think it's enough to link it here.

Reviewer 3: Fair enough. But one/two viz won't hurt I guess. I will let you choose those.

```

You have successfully augmented the individual images and prepared their annotations. However, preprocessing isn't
complete yet. In the final step, create a custom `collate_fn` to batch images together.
Pad images (which are now `pixel_values`) to the largest image in a batch, and create a corresponding `pixel_mask`
to indicate which pixels are real (1) and which are padding (0).

```py
>>> def collate_fn(batch):
...     pixel_values = [item["pixel_values"] for item in batch]
...     encoding = image_processor.pad_and_create_pixel_mask(pixel_values, return_tensors="pt")
...     labels = [item["labels"] for item in batch]
...     batch = {}
...     batch["pixel_values"] = encoding["pixel_values"]
...     batch["pixel_mask"] = encoding["pixel_mask"]
...     batch["labels"] = labels
...     return batch
```

> ## Training the DERT model

Reviewer 3: ```suggestion
## Training the DETR model
```

You have done most of the heavy lifting in the previous sections, so now you are ready to train your model!
The images in this dataset are still quite large, even after resizing. This means that finetuning this model will
require at least one GPU.

```py
>>> import torch

>>> device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

> ```

Reviewer 3: I think `trainer` will automatically take care of the device placements. No need to initialize this.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

Training involves the following steps:

> 1. Load the model with `DetrForObjectDetection.from_pretrained` using the same checkpoint as in the preprocessing.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

2. Define your training hyperparameters in `TrainingArguments`.
3. Pass the training arguments to `Trainer` along with the model, dataset, image processor, and data collator.

> 4. Call `train()`` to finetune your model.

Reviewer 3: ```suggestion
4. Call `train()` to finetune your model.
```

> 4. Call `train()`` to finetune your model.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

When loading the model from the same checkpoint that you used for the preprocessing, remember to pass the label to

> id and id to label maps that you created earlier from the dataset's metadata.

Reviewer 3: I think it's better to refer those maps as `label2id` and `id2label`, respectively.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

```py
>>> from transformers import DetrForObjectDetection

> >>> model = DetrForObjectDetection.from_pretrained(

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

...     checkpoint,
...     id2label=id2label,
...     label2id=label2id,
...     ignore_mismatched_sizes=True,
... )
```

> In the `TrainingArguments` use `output_dir` to specify where to save your model, then configure hyperparameters as you see fit.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

An important note: do not remove unused columns because this will drop the image column. Without the image column, you
can't create `pixel_values`. For this reason, set `remove_unused_columns` to `False`.
If you wish to share your model by pushing to the Hub, set `push_to_hub` to `True` (you must be signed in to Hugging
Face to upload your model).

```py
>>> from transformers import TrainingArguments

>>> training_args = TrainingArguments(
...     output_dir="detr-resnet-50_fine_tuned_cppe5",
...     per_device_train_batch_size=8,
...     num_train_epochs=10,
...     fp16=False,
...     save_steps=200,
...     logging_steps=50,
...     learning_rate=1e-5,
...     save_total_limit=2,
...     remove_unused_columns=False,
...     push_to_hub=True,
... )
```

Finally, bring everything together, and call `train`:

```py
>>> from transformers import Trainer

>>> trainer = Trainer(
...     model=model,
...     args=training_args,
...     data_collator=collate_fn,
...     train_dataset=cppe5["train"],
...     tokenizer=image_processor,
... )

> >>> rainer.train()

Reviewer 3: ```suggestion
>>> trainer.train()
```

```

## Evaluate
Object detection models are commonly evaluated with a set of [COCO-style metrics](https://cocodataset.org/#detection-eval).
You can use one of the existing metrics implementations, here you'll use the one from `torchvision`.

> Save your trained model to use for evaluation. Alternatively, you can evaluate the model that you have pushed to the Hub.

Reviewer 3: Same here. Not sure how the renderings aren't complete:

![image](https://user-images.githubusercontent.com/22957388/209902270-fd289ad8-f858-4822-8727-89dc9fbe3a6c.png)

Author: This is odd. Later in the text, the same link renders just fine. I'll try to fix it with markdown, but if that doesn't work, I'll replace it with HTML.

> Save your trained model to use for evaluation. Alternatively, you can evaluate the model that you have pushed to the Hub.

Reviewer 3: Note that we haven't yet called `trainer.push_to_hub()`, so I don't think the final model will be pushed to the Hub yet.

Author: Ooh, that's actually very subtle! I noticed that it was pushed to hub at checkpoints when we have `push_to_hub=True,` in the training arguments, and I assumed that it was pushing the final model too. If that's not the case, I definitely should add `trainer.push_to_hub()` after evaluation.

Reviewer 3: Additionally, `trainer.push_to_hub()` will create the README for the model repository.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

```py

> >>> trainer.save_model("./detr-resnet-50_fine_tuned_cppe5")

Author: Should it be here or after we run the evaluation? What if I'm not happy with the evaluation result, and want to train some more?

> >>> trainer.save_model("./detr-resnet-50_fine_tuned_cppe5")

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

```

To use the `torchvision` evaluator, you'll need to prepare a ground truth COCO dataset. The API to build a COCO dataset
requires the data to be stored in a certain format, so you'll need to save images and annotations to disk first. Just like
when you prepared your data for training, the annotations from the `cppe5["test"]` need to be formatted. However, images
should stay as they are.

The following example shows how you can:
- Create a torchvision CocoDetection dataset
- Reformat annotations from the `cppe5["test"]`
- Save images and annotations into files that you can create an instance of CocoDetection dataset from
- Use `cocoevaluate` metric to compute the [COCO metrics](https://cocodataset.org/#detection-eval).

> ```py

Reviewer 3: This code snippet is a bit long. 

I would break it down into pieces and places connectors in between them. 

* Preparing `CocoDetection` dataset for compatibility.
* Serializing images. 
* Evaluation. 

Also consider a different name than `CocoDetection` since it might confuse the readers. We aren't actually evaluating on COCO here.

Reviewer: This is indeed way too long for this guide. It should be moved to a more advanced guide or juse be part in an example (this whole guide would make a great example once converted to a Python script) that we can link here.

Author: Phew, the evaluation part is long indeed. Unfortunately, there's no simpler way (yet) to properly use the coco metrics. I can split this into several pieces and rename parts so that it isn't as confusing.

Reviewer 3: Yeah let's start with that.

Author: Split into parts and added some explanation between the blocks (in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7), however, I may need to do another pass at the evaluation part. @younesbelkada is working on optimizing this.

import os
import json
import evaluate
import torchvision
import numpy as np
from tqdm import tqdm

> class CocoDetection(torchvision.datasets.CocoDetection):

Author: From what I understand, COCO API is very strict in what it requires. It expects to have the images saved, and the annotations to be stored on a json file. @younesbelkada was very helpful with the evaluation part and he has more insight on the inner workings of the COCO API and why we need CocoDetection dataset.

    def __init__(self, img_folder, feature_extractor, ann_file):

>         super(CocoDetection, self).__init__(img_folder, ann_file)

Reviewer 3: I forgot the exact Python version but since that version we can do:

```py
        super().__init__(img_folder, ann_file)
```

Author: Neat!

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

        self.feature_extractor = feature_extractor

    def __getitem__(self, idx):
        # read in PIL image and target in COCO format
        img, target = super(CocoDetection, self).__getitem__(idx)

>         # preprocess image and target (converting target to DETR format, resizing + normalization of both image and target)

Reviewer 3: Break this comment into two lines. It will read easier.

        image_id = self.ids[idx]
        target = {"image_id": image_id, "annotations": target}
        encoding = self.feature_extractor(images=img, annotations=target, return_tensors="pt")
        pixel_values = encoding["pixel_values"].squeeze()  # remove batch dimension
        target = encoding["labels"][0]  # remove batch dimension

        return {"pixel_values": pixel_values, "labels": target}

im_processor = DetrImageProcessor.from_pretrained("./detr-resnet-50_fine_tuned_cppe5")

> model = DetrForObjectDetection.from_pretrained("./detr-resnet-50_fine_tuned_cppe5")

Reviewer 3: We usually load the fine-tuned models from the Hub. But I don't have strong opinions about it.

Author: I will probably switch to loading from the Hub. Then I won't need to save locally either, and it'll be slightly more consistent. My initial thought was that I might like to evaluate first before sharing. But it doesn't change that much here.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

> model = DetrForObjectDetection.from_pretrained("./detr-resnet-50_fine_tuned_cppe5")

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

# prepare the test dataset
def val_formatted_anns(image_id, objects):
    annotations = []
    for i in range(0, len(objects["id"])):
        new_ann = {
            "id": objects["id"][i],
            "category_id": objects["category"][i],
            "iscrowd": 0,
            "image_id": image_id,
            "area": objects["area"][i],
            "bbox": objects["bbox"][i],
        }
        annotations.append(new_ann)

    return annotations

# CocoDetection expects to read the data from files in a certain format

> def save_cppe5_annotation_file_images(cppe5):

Reviewer 3: Since this snippet is long, consider grouping semantically close lines of code together and putting comments on top of them.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

    output_json = {}
    path_output_cppe5 = f"{os.getcwd()}/cppe5/"

    if not os.path.exists(path_output_cppe5):
        os.makedirs(path_output_cppe5)

    path_anno = os.path.join(path_output_cppe5, "cppe5_ann.json")
    categories_json = [{"supercategory": "none", "id": id, "name": id2label[id]} for id in id2label]
    output_json["images"] = []
    output_json["annotations"] = []

>     for batch in cppe5:

Reviewer 3: Are we getting batches of samples or individual samples? Based on that `batch` might have to be renamed.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

        ann = val_formatted_anns(batch["image_id"], batch["objects"])
        output_json["images"].append(
            {
                "id": batch["image_id"],
                "width": batch["image"].width,
                "height": batch["image"].height,
                "file_name": f"{batch['image_id']}.png",
            }
        )
        output_json["annotations"].extend(ann)
    output_json["categories"] = categories_json

    with open(path_anno, "w") as file:
        json.dump(output_json, file, ensure_ascii=False, indent=4)

    for im, img_id in zip(cppe5["image"], cppe5["image_id"]):
        path_img = os.path.join(path_output_cppe5, f"{img_id}.png")
        im.save(path_img)

    return path_output_cppe5, path_anno

path_output_cppe5, path_anno = save_cppe5_annotation_file_images(cppe5["test"])

> dummy = CocoDetection(path_output_cppe5, im_processor, path_anno)

Reviewer 3: I don't think it's a dummy dataset. 

Suggestion: `test_ds_coco_format`.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

module = evaluate.load("ybelkada/cocoevaluate", coco=dummy.coco)
val_dataloader = torch.utils.data.DataLoader(dummy, batch_size=8, shuffle=False, num_workers=4, collate_fn=collate_fn)

with torch.no_grad():
    for idx, batch in enumerate(tqdm(val_dataloader)):
        pixel_values = batch["pixel_values"]
        pixel_mask = batch["pixel_mask"]

        labels = [
            {k: v for k, v in t.items()} for t in batch["labels"]
        ]  # these are in DETR format, resized + normalized

        # forward pass
        outputs = model(pixel_values=pixel_values, pixel_mask=pixel_mask)

        orig_target_sizes = torch.stack([target["orig_size"] for target in labels], dim=0)
        results = im_processor.post_process(outputs, orig_target_sizes)  # convert outputs of model to COCO api

        module.add(prediction=results, reference=labels)
        del batch

results = module.compute()

> print(results)

Reviewer 3: Spend a few sentences discussing the results and also append the output of `print(results)`.

```

## Inference

> Now that you have finetuned a DETR model, evaluated it and uploaded it to the 🤗Hub, you can use it for inference.

Reviewer 3: ```suggestion
Now that you have finetuned a DETR model, evaluated it, and uploaded it to the Hugging Face Hub, you can use it for inference.
```

Reviewer 3: When mentioning Hub for the first time, we do "Hugging Face Hub" and then we just refer to it as "Hub". 

This a slight deviation to how we mention the other ecosystem tools (which usually have a 🤗 prefix).

```py
>>> import requests

> >>> url = "https://chartwell.com/-/media/Images/blog/2020/04/keeping-staff-and-residents-protected.jpeg?mw=1900&hash=B8A76374D6BD24B0DC8A779D0C985027"

Reviewer 3: Feel free to cut the URL to:

```
https://chartwell.com/-/media/Images/blog/2020/04/keeping-staff-and-residents-protected.jpeg
```

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

>>> image = Image.open(requests.get(url, stream=True).raw)

> >>> image_processor = DetrImageProcessor.from_pretrained("MariaK/detr-resnet-50_fine_tuned_cppe5")

Reviewer: The first example should just use the `pipeline` function, which will take care of all pre/post processing. Then you can dive into doing it by hand.

Author: Added an example of inference with the pipeline. Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

> >>> model = DetrForObjectDetection.from_pretrained("MariaK/detr-resnet-50_fine_tuned_cppe5")

Reviewer: Use the Auto API here.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

>>> inputs = image_processor(images=image, return_tensors="pt")

> >>> outputs = model(**inputs)

Reviewer 3: This needs to be placed under `with torch.no_grad():` context.

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

> >>> # convert outputs (bounding boxes and class logits) to COCO API

Reviewer: There shouldn't be comments in such a guide: just place text between code blocks :-)

Author: Fixed in 1a38260b27ef977e0d82ba1c918eeb19c8f3dcb7

>>> target_sizes = torch.tensor([image.size[::-1]])
>>> results = image_processor.post_process_object_detection(outputs, threshold=0.9, target_sizes=target_sizes)[0]

>>> for score, label, box in zip(results["scores"], results["labels"], results["boxes"]):
...     box = [round(i, 2) for i in box.tolist()]
...     print(
...         f"Detected {model.config.id2label[label.item()]} with confidence "
...         f"{round(score.item(), 3)} at location {box}"
...     )
```

> 

Reviewer 3: Needs to be deleted.

Reviewer 3: We also need to show the results (outputs of print) and visualize the results.

annotations identifying medical personal protective equipment (PPE) in the context of the COVID-19 pandemic.

Start by loading the dataset:

```py
>>> from datasets import load_dataset

>>> cppe5 = load_dataset("cppe-5")
>>> cppe5
DatasetDict({
    train: Dataset({
        features: ['image_id', 'image', 'width', 'height', 'objects'],
        num_rows: 1000
    })
    test: Dataset({
        features: ['image_id', 'image', 'width', 'height', 'objects'],
        num_rows: 29
    })
})
```

You'll see that this dataset already comes with a training set containing 1000 images and a test set with 29 images.

To get familiar with the data, explore what the examples look like.

```py
>>> cppe5["train"][0]
{'image_id': 15,
 'image': <PIL.JpegImagePlugin.JpegImageFile image mode=RGB size=943x663 at 0x7F9EC9E77C10>,
 'width': 943,
 'height': 663,
 'objects': {'id': [114, 115, 116, 117],
  'area': [3796, 1596, 152768, 81002],
  'bbox': [[302.0, 109.0, 73.0, 52.0],
   [810.0, 100.0, 57.0, 28.0],
   [160.0, 31.0, 248.0, 616.0],
   [741.0, 68.0, 202.0, 401.0]],
  'category': [4, 4, 0, 0]}}
```

The examples in the dataset have the following fields:

> - `image_id`: the example image id

Author: Oh, thanks for noticing! I think it's actually better to remove the dots from the `objects` because these aren't full sentences.

> - `image_id`: the example image id

Author: Fixed in aed4b71771a47d3fe31b233d0b89f3f7314a0002

...     if not os.path.exists(path_output_cppe5):
...         os.makedirs(path_output_cppe5)

...     path_anno = os.path.join(path_output_cppe5, "cppe5_ann.json")
...     categories_json = [{"supercategory": "none", "id": id, "name": id2label[id]} for id in id2label]
...     output_json["images"] = []
...     output_json["annotations"] = []
...     for example in cppe5:
...         ann = val_formatted_anns(example["image_id"], example["objects"])
...         output_json["images"].append(
...             {
...                 "id": example["image_id"],
...                 "width": example["image"].width,
...                 "height": example["image"].height,
...                 "file_name": f"{example['image_id']}.png",
...             }
...         )
...         output_json["annotations"].extend(ann)
...     output_json["categories"] = categories_json

...     with open(path_anno, "w") as file:
...         json.dump(output_json, file, ensure_ascii=False, indent=4)

...     for im, img_id in zip(cppe5["image"], cppe5["image_id"]):
...         path_img = os.path.join(path_output_cppe5, f"{img_id}.png")
...         im.save(path_img)

...     return path_output_cppe5, path_anno
```

> Next, prepare an instance of a CocoDetection class that can be used with `cocoevaluator`.

Reviewer 2: ```suggestion
Next, prepare an instance of a `CocoDetection` class that can be used with `cocoevaluator`.
```

...         module.add(prediction=results, reference=labels)
...         del batch

>>> results = module.compute()
>>> print(results)
Accumulating evaluation results...
DONE (t=0.08s).
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.150
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.280
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.130
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.038
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.036
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.182
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.166
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.317
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.335
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.104
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.146
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.382
```

> These results can be further improved by adjusting the hyperparameters in the `training_args`. Give it a go!

Reviewer 2: ```suggestion
These results can be further improved by adjusting the hyperparameters in [`~transformers.TrainingArguments`]. Give it a go!
```

## Inference
Now that you have finetuned a DETR model, evaluated it, and uploaded it to the Hugging Face Hub, you can use it for inference.

> The simplest way to try out your finetuned model for inference is to use it in a `pipeline()`. Instantiate a pipeline

Reviewer 2: ```suggestion
The simplest way to try out your finetuned model for inference is to use it in a [`Pipeline`]. Instantiate a pipeline
```

an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
specific language governing permissions and limitations under the License.
-->

# Object detection

[[open-in-colab]]

Object detection is a computer vision task of detecting instances of semantic objects that belong to a particular
class (such as humans, buildings, or cars) in parts of an image. Object detection models receive an image as input and output
coordinates of the bounding boxes and associated labels of the detected objects. An image can contain multiple objects,
each with its own bounding box and a label (e.g. it can have a car and a building).
be present in different parts of an image (e.g. the image can have several cars).
This task is commonly used in autonomous driving for detecting things like pedestrians, road signs, and traffic lights.
Other applications include counting objects in images, image search, and more.

 <Tip>
Check out the <a href="https://huggingface.co/tasks/object-detection">object detection</a> task page to learn about use cases,
models, metrics, and datasets associated with this task.
 </Tip>

In this guide, you will learn how to:

>  1. Finetune [DETR](https://huggingface.co/docs/transformers/model_doc/detr), a model that combines a convolutional

Reviewer 4: ```suggestion
 1. Fine-tune [DETR](https://huggingface.co/docs/transformers/model_doc/detr), a model that combines a convolutional
```

Reviewer: In an opensource repo like Transformers, it's going to be very tedious to maintain a consistent spelling since we have so many contributors (for this or pre-trained/pretrained). The positive impact of forcing one spelling is exactly zero since both are perfectly understandable, while taking time from reviewers/maintainers, so overall a negative impact.

 backbone with an encoder-decoder Transformer, on the [CPPE-5](https://huggingface.co/datasets/cppe-5)
 dataset.
 2. Use your finetuned model for inference.

Before you begin, make sure you have all the necessary libraries installed:

```bash
pip install -q datasets transformers evaluate timm albumentations
```

In this guide you'll use `datasets` to load a dataset from the Hugging Face Hub, `transformers` to train your model,
and `albumentations` to augment the data. `timm` is currently required to load a convolutional backbone for the DETR model.

We encourage you to share your model with the community. Log in to your Hugging Face account to upload it to Hub.
When prompted, enter your token to log in:

```py
>>> from huggingface_hub import notebook_login

>>> notebook_login()
```

## Load the CPPE-5 dataset

In this guide, you'll use the [CPPE-5 dataset](https://huggingface.co/datasets/cppe-5) that contains images with

> annotations identifying medical personal protective equipment (PPE) in the context of the COVID-19 pandemic.

Reviewer 4: Please include a guide on how people can use their own data. Using a HuggingFace dataset is nice, but the first question people always come up with is "but how I can use my own custom dataset?".

Here I'd recommend adding a link to ImageFolder + metadata, which allows people to turn their own image captions into a 🤗 Dataset: https://huggingface.co/docs/datasets/v2.8.0/en/image_dataset#image-captioning

Reviewer: This is out of scope of the tasks guide. None of the other one does that.

Author: If this is a common question, perhaps, this can be a note or a tip in this doc? https://huggingface.co/docs/transformers/training
Something along the lines "In the Transformers documentation we use datasets from the Hugging Face Hub. If you would like to use your own custom data, check out the Datasets documentation on how to create a dataset" ?

Reviewer 3: That would be a good addition IMO.

```

<div class="flex justify-center">
    <img src="https://i.imgur.com/TdaqPJO.png" alt="CPPE-5 Image Example"/>
</div>

To visualize the bounding boxes with associated labels, you can get the labels from the dataset's metadata, specifically
the `category` field.
You'll also want to create dictionaries that map a label id to a label class (`id2label`) and the other way around (`label2id`).
You can use them later when setting up the model. Including these maps will make your model reusable by others if you share
it on the Hugging Face Hub.

As a final step of getting familiar with the data, explore it for potential issues. One common problem with datasets for
object detection is bounding boxes that "stretch" beyond the edge of the image. Such "runaway" bounding boxes can raise
errors during training and should be addressed at this stage. There are a few examples with this issue in this dataset.
To keep things simple in this guide, we remove these images from the data.

```py
>>> remove_idx = [590, 821, 822, 875, 876, 878, 879]
>>> keep = [i for i in range(len(cppe5["train"])) if i not in remove_idx]
>>> cppe5["train"] = cppe5["train"].select(keep)
```

## Preprocess the data

> To finetune a model, you must preprocess the data you plan to use to match precisely the approach used for the pre-trained model.

Reviewer 4: ```suggestion
To fine-tune a model, you must preprocess the data you plan to use to match precisely the approach used for the pre-trained model.
```

[`AutoImageProcessor`] takes care of processing image data to create `pixel_values`, `pixel_mask`, and
`labels` that a DETR model can train with. The image processor has some attributes that you won't have to worry about:

- `image_mean = [0.485, 0.456, 0.406 ]`
- `image_std = [0.229, 0.224, 0.225]`

These are the mean and standard deviation used to normalize images during the model pre-training. These values are crucial
to replicate when doing inference or finetuning a pre-trained image model.

> Instantiate the image processor from the same checkpoint as the model you want to finetune.

Reviewer 4: ```suggestion
Instantiate the image processor from the same checkpoint as the model you want to fine-tune.
```

```py
>>> from transformers import AutoImageProcessor

>>> checkpoint = "facebook/detr-resnet-50"
>>> image_processor = AutoImageProcessor.from_pretrained(checkpoint)
```

Before passing the images to the `image_processor`, apply two preprocessing transformations to the dataset:
- Augmenting images
- Re-formatting annotations to meet DETR expectations

> First, to make sure the model does not overfit on the training data, you can apply image augmentation with [Albumentations](https://albumentations.ai/docs/).

Reviewer 4: ```suggestion
First, to make sure the model does not overfit on the training data, you can apply image augmentation with any data augmentation library. Here we use [Albumentations](https://albumentations.ai/docs/) ...
```

to indicate which pixels are real (1) and which are padding (0).

```py
>>> def collate_fn(batch):
...     pixel_values = [item["pixel_values"] for item in batch]
...     encoding = image_processor.pad_and_create_pixel_mask(pixel_values, return_tensors="pt")
...     labels = [item["labels"] for item in batch]
...     batch = {}
...     batch["pixel_values"] = encoding["pixel_values"]
...     batch["pixel_mask"] = encoding["pixel_mask"]
...     batch["labels"] = labels
...     return batch
```

## Training the DETR model
You have done most of the heavy lifting in the previous sections, so now you are ready to train your model!
The images in this dataset are still quite large, even after resizing. This means that finetuning this model will
require at least one GPU.

Training involves the following steps:
1. Load the model with [`AutoModelForObjectDetection`] using the same checkpoint as in the preprocessing.
2. Define your training hyperparameters in [`TrainingArguments`].
3. Pass the training arguments to [`Trainer`] along with the model, dataset, image processor, and data collator.

> 4. Call [`~Trainer.train`] to finetune your model.

Reviewer 4: ```suggestion
4. Call [`~Trainer.train`] to fine-tune your model.
```

> When loading the model from the same checkpoint that you used for the preprocessing, remember to pass the `label2id`
> and `id2label` maps that you created earlier from the dataset's metadata.

Reviewer 4: ```suggestion
When loading the model from the same checkpoint that you used for the preprocessing, remember to pass the `label2id`
and `id2label` maps that you created earlier from the dataset's metadata. Additionally, we specify `ignore_mismatched_sizes=True` to replace the existing classification head with a new one.
```

```py
>>> from transformers import AutoModelForObjectDetection

>>> model = AutoModelForObjectDetection.from_pretrained(
...     checkpoint,
...     id2label=id2label,
...     label2id=label2id,

> ...     ignore_mismatched_sizes=True,

Reviewer 4: Please add a sentence to explain why we set `ignore_mismatched_sizes` here

>>> print(results)
Accumulating evaluation results...
DONE (t=0.08s).
IoU metric: bbox
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.150
 Average Precision  (AP) @[ IoU=0.50      | area=   all | maxDets=100 ] = 0.280
 Average Precision  (AP) @[ IoU=0.75      | area=   all | maxDets=100 ] = 0.130
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.038
 Average Precision  (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.036
 Average Precision  (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.182
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=  1 ] = 0.166
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets= 10 ] = 0.317
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=   all | maxDets=100 ] = 0.335
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.104
 Average Recall     (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.146
 Average Recall     (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.382
```
These results can be further improved by adjusting the hyperparameters in the `training_args`. Give it a go!

## Inference

> Now that you have finetuned a DETR model, evaluated it, and uploaded it to the Hugging Face Hub, you can use it for inference.
> The simplest way to try out your finetuned model for inference is to use it in a `pipeline()`. Instantiate a pipeline
> for object detection with your model, and pass an image to it:

Reviewer 4: ```suggestion
Now that you have fine-tuned a DETR model, evaluated it, and uploaded it to the Hugging Face Hub, you can use it for inference.
The simplest way to try out your fine-tuned model for inference is to use it in a `pipeline()`. Instantiate a pipeline
for object detection with your model, and pass an image to it:
```
